File: Python/202.Happy Number.py
#Happy Number
#一直循环做：把数的各个位数求平方和。直到结果为1，就是Happy Number。
class Solution(object):
    def isHappy(self, n):
        dic = {}
        while n!=1 and (dic.get(n,-1))==-1:
            dic[n] = 1
            l = [int(x) for x in str(n)]
            n = sum(map(self.square,l))
        if n==1:
            return True
        else:
            return False

    def square(self,x):
        return x*x


s = Solution()
print(s.isHappy(21))

File: C++/Math/202.Happy Number.cpp
/*
HappyNumber，给一个多位数，循环做每位数平方相加，能得到结果1，则是happynumber。
*/

#include <iostream>
#include <vector>
#include <unordered_set>
#include <cmath>
using namespace std;

class Solution {
public:

	//发现2--6都是非happynumber，所有的非happynumber最终会进入这个循环。
	bool isHappy(int n) 
	{
		while(n>6)
		{
        	int next = 0;
        	while(n){next+=(n%10)*(n%10); n/=10;}
        	n = next;
    	}
    	return n==1;
	}

	/*使用set
	bool isHappy(int n) {
		unordered_set<int>	myset;
        while(1)
        {
        	int sum = 0,tmp = n;
        	while(tmp>0)
        	{
        		sum += (tmp%10)*(tmp%10);
        		tmp = tmp/10;
        	}
        	if(sum==1)	return 1;
        	if (myset.find(sum)!=myset.end())	return false;
			else
				myset.insert(sum);
			n = sum;
        }
    }*/
};

int main()
{
	Solution s;
	cout<<s.isHappy(82)<<endl;
}

